src/semantic-highlighting.ts

```diff
@@ -0,0 +1,167 @@
+  tokens: string;
+}
+
+// Parameters for the semantic highlighting (server-side) push notification.
+// Mirrors the structure in the semantic highlighting proposal for LSP.
+interface SemanticHighlightingParams {
+  // The text document that has to be decorated with the semantic highlighting
+  // information.
+  textDocument: VersionedTextDocumentIdentifier;
+  // An array of semantic highlighting information.
+  lines: SemanticHighlightingInformation[];
+}
+
+// A SemanticHighlightingToken decoded from the base64 data sent by clangd.
+interface SemanticHighlightingToken {
+  // Start column for this token.
+  character: number;
+  // Length of the token.
+  length: number;
+  // The TextMate scope index to the clangd scope lookup table.
+  scopeIndex: number;
+  kind: string;
+}
+
+// A line of decoded highlightings from the data clangd sent.
+export interface SemanticHighlightingLine {
+  // The zero-based line position in the text document.
+  line: number;
+  // All SemanticHighlightingTokens on the line.
+  tokens: SemanticHighlightingToken[];
+}
+
+export class SemanticHighlightingFeature implements StaticFeature {
+  private scopeTable: string[][] = [];
+
+  constructor(client: BaseLanguageClient, context: ExtensionContext) {
+    context.subscriptions.push(
+      client.onDidChangeState(({ newState }) => {
+        if (newState === State.Running) {
+          const notification = new NotificationType<SemanticHighlightingParams, void>('textDocument/semanticHighlighting');
+          client.onNotification(notification, this.handleNotification.bind(this));
+        }
+      })
+    );
+  }
+
+  initialize(capabilities: ServerCapabilities) {
+    const serverCapabilities: ServerCapabilities & { semanticHighlighting?: { scopes: string[][] } } = capabilities;
+    if (!serverCapabilities.semanticHighlighting) return;
+
+    this.scopeTable = serverCapabilities.semanticHighlighting.scopes;
+  }
+
+  fillClientCapabilities(capabilities: ClientCapabilities) {
+    const textDocumentCapabilities: TextDocumentClientCapabilities & { semanticHighlightingCapabilities?: { semanticHighlighting: boolean } } = capabilities.textDocument!;
+    textDocumentCapabilities.semanticHighlightingCapabilities = {
+      semanticHighlighting: true,
+    };
+  }
+
+  async handleNotification(params: SemanticHighlightingParams) {
+    // use https://github.com/jackguo380/vim-lsp-cxx-highlight to do highlighting
+    const lines: SemanticHighlightingLine[] = params.lines.map((line) => ({ line: line.line, tokens: this.decodeTokens(line.tokens) }));
+    const symbols: any[] = [];
+    const skipped: Range[] = [];
+
+    for (const line of lines) {
+      for (const token of line.tokens) {
+        if (token.kind === 'InactiveCode') {
+          skipped.push(Range.create(line.line, token.character, line.line, token.character + token.length));
+        } else {
+          symbols.push({
+            id: 0,
+            kind: token.kind,
```

Comment:
I'm kind of confused how this works currently since `kind` `parentKind` and `storage` are supposed to be integer values.

See this message from ccls:
```
{
   "jsonrpc":"2.0",
   "method":"$ccls/publishSemanticHighlight",
   "params":{
      "uri":"file:///home/guoj/.vim/ccls/src/test.hh",
      "symbols":[
         {
            "id":43742,
            "parentKind":0,
            "kind":12,
            "storage":0,
            "ranges":[ ],
            "lsRanges":[ { "start":{ "line":8, "character":5 }, "end":{ "line":8, "character":18 } } ]
         },
         {
            "id":0,
            "parentKind":0,
            "kind":3,
            "storage":0,
            "ranges":[ ],
            "lsRanges":[ { "start":{ "line":8, "character":25 }, "end":{ "line":8, "character":28 } } ]
         },
         {
            "id":2,
            "parentKind":0,
            "kind":3,
            "storage":0,
            "ranges":[ ],
            "lsRanges":[ { "start":{ "line":7, "character":10 }, "end":{ "line":7, "character":14 } } ]
         }
      ]
   }
}
```

Reply:
Okay I decided to resolves this on vim-lsp-cxx-highlight's end since the mapping of groups to indices is kind of a nuisance and makes adding new highlight groups a pain.

jackguo380/vim-lsp-cxx-highlight@6b8eef2da

Reply:
Don't worry, I've already parse kind. Currently, ccls/cquery returns integer value of kind, then vim-lsp-cxx-highlight will handle them with `lsp_cxx_hl#notify_symbols`, then `lsp_cxx_hl#parse#normalize_symbols` to parse integer to string value, and finally `lsp_cxx_hl#hl#notify_symbols` to do highlighting. coc-clangd already parse these, and call `lsp_cxx_hl#hl#notify_symbols` directly.